File: internal/query/filter.go
package query

import (
	"fmt"

	"vitess.io/vitess/go/vt/sqlparser"
)

// sqlParser is a package-level parser instance used for filter expression validation
// and column rewriting. Using sqlparser.New instead of NewTestParser for production use.
var sqlParser *sqlparser.Parser

func init() {
	var err error
	sqlParser, err = sqlparser.New(sqlparser.Options{})
	if err != nil {
		panic(fmt.Sprintf("failed to create SQL parser: %v", err))
	}
}

// ValidateFilterExpression parses a WHERE clause expression into an AST and
// validates that it contains only safe node types (comparisons, boolean logic,
// column refs, and literals). Rejects subqueries, function calls, and other
// potentially dangerous constructs.
func ValidateFilterExpression(expr string) error {
	if expr == "" {
		return nil
	}

	// Parse as a SELECT WHERE clause so vitess can produce an AST.
	stmt, err := sqlParser.Parse("SELECT 1 FROM t WHERE " + expr)
	if err != nil {
		return fmt.Errorf("invalid filter expression: %w", err)
	}

	sel, ok := stmt.(*sqlparser.Select)
	if !ok || sel.Where == nil {
		return fmt.Errorf("invalid filter expression")
	}

	return walkExpr(sel.Where.Expr)
}

// walkExpr recursively validates that an expression node is safe.
func walkExpr(node sqlparser.Expr) error {
	switch n := node.(type) {
	case *sqlparser.AndExpr:
		if err := walkExpr(n.Left); err != nil {
			return err
		}
		return walkExpr(n.Right)

	case *sqlparser.OrExpr:
		if err := walkExpr(n.Left); err != nil {
			return err
		}
		return walkExpr(n.Right)

	case *sqlparser.NotExpr:
		return walkExpr(n.Expr)

	case *sqlparser.ComparisonExpr:
		// Allow: =, !=, <, >, <=, >=, <=>, LIKE, NOT LIKE, IN, NOT IN, REGEXP
		if err := walkExpr(n.Left); err != nil {
			return err
		}
		if err := walkExpr(n.Right); err != nil {
			return err
		}
		if n.Escape != nil {
			if err := walkExpr(n.Escape); err != nil {
				return err
			}
		}
		return nil

	case *sqlparser.BetweenExpr:
		// BETWEEN ... AND ...
		if err := walkExpr(n.Left); err != nil {
			return err
		}
		if err := walkExpr(n.From); err != nil {
			return err
		}
		return walkExpr(n.To)

	case *sqlparser.IsExpr:
		// IS NULL, IS NOT NULL, IS TRUE, IS FALSE
		return walkExpr(n.Left)

	case *sqlparser.ColName:
		// Column references are always safe.
		return nil

	case *sqlparser.Literal:
		// String, int, float, hex literals are safe.
		return nil

	case sqlparser.BoolVal:
		return nil

	case sqlparser.ValTuple:
		// Tuple of values, e.g. IN (1, 2, 3)
		for _, v := range n {
			if err := walkExpr(v); err != nil {
				return err
			}
		}
		return nil

	case *sqlparser.NullVal:
		return nil

	case *sqlparser.UnaryExpr:
		// Unary minus for negative numbers.
		return walkExpr(n.Expr)

	default:
		return fmt.Errorf("unsupported expression type %T in filter", node)
	}
}

File: crates/metalog-query/src/filter.rs
use sqlparser::{
    ast::{Expr, Statement, UnaryOperator},
    dialect::MySqlDialect,
    parser::Parser,
};

/// Validates a filter expression for safety.
///
/// Parses the expression as `SELECT 1 FROM t WHERE <expr>` and walks the AST,
/// allowing only safe constructs: comparisons, AND/OR/NOT, LIKE, IN, BETWEEN,
/// IS NULL/IS NOT NULL, column references, and literals.
///
/// Rejects: subqueries, function calls, UNION, JOIN, etc.
pub fn validate_filter_expression(expr: &str) -> Result<(), FilterError> {
    if expr.is_empty() {
        return Ok(());
    }

    let sql = format!("SELECT 1 FROM t WHERE {expr}");
    let dialect = MySqlDialect {};
    let statements =
        Parser::parse_sql(&dialect, &sql).map_err(|e| FilterError::Parse(e.to_string()))?;

    let stmt = statements
        .first()
        .ok_or_else(|| FilterError::Parse("empty parse result".into()))?;

    let Statement::Query(query) = stmt else {
        return Err(FilterError::Unsafe("expected SELECT statement".into()));
    };

    let sqlparser::ast::SetExpr::Select(select) = query.body.as_ref() else {
        return Err(FilterError::Unsafe("expected simple SELECT".into()));
    };

    let Some(ref selection) = select.selection else {
        return Err(FilterError::Unsafe("missing WHERE clause".into()));
    };

    validate_expr(selection)?;
    Ok(())
}

fn validate_expr(expr: &Expr) -> Result<(), FilterError> {
    match expr {
        // Safe leaf nodes.
        Expr::Identifier(_) | Expr::CompoundIdentifier(_) => Ok(()),
        Expr::Value(_) => Ok(()),

        // Binary operations (comparisons, AND, OR).
        Expr::BinaryOp { left, right, .. } => {
            validate_expr(left)?;
            validate_expr(right)
        }

        // Unary NOT.
        Expr::UnaryOp {
            op: UnaryOperator::Not,
            expr: inner,
        } => validate_expr(inner),

        // IS NULL / IS NOT NULL.
        Expr::IsNull(inner) | Expr::IsNotNull(inner) => validate_expr(inner),

        // IN (value list).
        Expr::InList {
            expr: inner, list, ..
        } => {
            validate_expr(inner)?;
            for item in list {
                validate_expr(item)?;
            }
            Ok(())
        }

        // BETWEEN.
        Expr::Between {
            expr: inner,
            low,
            high,
            ..
        } => {
            validate_expr(inner)?;
            validate_expr(low)?;
            validate_expr(high)
        }

        // LIKE.
        Expr::Like {
            expr: inner,
            pattern,
            ..
        } => {
            validate_expr(inner)?;
            validate_expr(pattern)
        }

        // Nested parentheses.
        Expr::Nested(inner) => validate_expr(inner),

        // Everything else is rejected.
        other => Err(FilterError::Unsafe(format!(
            "unsupported expression type: {other}"
        ))),
    }
}

#[derive(Debug, thiserror::Error)]
pub enum FilterError {
    #[error("filter parse error: {0}")]
    Parse(String),

    #[error("unsafe filter expression: {0}")]
    Unsafe(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_expressions() {
        let cases = [
            "min_timestamp > 1000",
            "min_timestamp >= 1000 AND max_timestamp < 2000",
            "state = 'IR_CLOSED'",
            "min_timestamp BETWEEN 1000 AND 2000",
            "state IN ('IR_CLOSED', 'ARCHIVE_CLOSED')",
            "(min_timestamp > 1000) AND (state = 'IR_CLOSED')",
            "NOT state = 'IR_PURGING'",
            "clp_ir_path IS NOT NULL",
            "clp_archive_path IS NULL",
            "state LIKE 'IR_%'",
        ];
        for expr in cases {
            assert!(
                validate_filter_expression(expr).is_ok(),
                "should be valid: {expr}"
            );
        }
    }

    #[test]
    fn empty_expression() {
        assert!(validate_filter_expression("").is_ok());
    }

    #[test]
    fn reject_subquery() {
        let result = validate_filter_expression("id IN (SELECT id FROM other)");
        assert!(result.is_err());
    }

    #[test]
    fn reject_function_call() {
        let result = validate_filter_expression("NOW() > min_timestamp");
        assert!(result.is_err());
    }

    #[test]
    fn reject_invalid_sql() {
        let result = validate_filter_expression("this is not sql at all");
        assert!(result.is_err());
    }
}
